### PR TITLE
Issue 7562 22

### PR DESCRIPTION
**Pull request by laa**
Please review fix for issue #7562.

### CODE REVIEW

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java [hunk 1]
```diff
@@ -817,4 +910,35 @@ public void run() {
     }
   }
 
+  private static void readByteBuffer(ByteBuffer buffer, FileChannel channel, long position, boolean throwOnEof) throws IOException {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: Seems like copy-pasted from `OFileClassic` ;)

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java [hunk 2]
```diff
@@ -817,4 +910,35 @@ public void run() {
...
+        else {
+          buffer.put(new byte[buffer.remaining()]);
+          return;
+        }
+
+      read += r;
+    }
+  }
+
+  private static void writeByteBuffer(ByteBuffer buffer, FileChannel channel, long position) throws IOException {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: Same as above.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import com.orientechnologies.common.serialization.types.OIntegerSerializer;
 import com.orientechnologies.common.serialization.types.OLongSerializer;
 import com.orientechnologies.orient.core.config.OGlobalConfiguration;
+import com.orientechnologies.orient.core.sql.parser.OInteger;
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: Irrelevant import.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java [hunk 4]
```diff
@@ -30,10 +31,29 @@
...
-  public static final  int RECORDS_OFFSET      = FREE_SPACE_OFFSET + OIntegerSerializer.INT_SIZE;
+
+  /**
+   * Information about position of LSN of last record which end is stored on this page. If only begging of the record is stored on
+   * this page its position is not stored.
+   *
+   * @see OLogSegment.FlushTask#commitLog()
+   * @see OLogSegment#init(ByteBuffer)
+   */
+  public static final int LAST_STORED_LSN = FREE_SPACE_OFFSET + OIntegerSerializer.INT_SIZE;
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: Seems like the layout of the WAL page is changed. Will it be safe/possible to use the new version on the old WAL?

File: core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java [hunk 5]
```diff
@@ -178,9 +178,11 @@ public void change(Object currentValue, Object newValue) {
   STORAGE_MAKE_FULL_CHECKPOINT_AFTER_CREATE("storage.makeFullCheckpointAfterCreate",
       "Indicates whether a full checkpoint should be performed, if storage was created", Boolean.class, false),
 
+  @Deprecated
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java]: fixe

File: core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java [hunk 6]
```diff
@@ -178,9 +178,11 @@ public void change(Object currentValue, Object newValue) {
...
+
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java (hunk 3 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java (hunk 4 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: yes, you are right added change which will ensure backward compatibility

File: core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java (hunk 5 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java (hunk 6 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java [hunk 7]
```diff
@@ -371,11 +413,100 @@ public long getOrder() {
   }
 
   public void init(ByteBuffer buffer) throws IOException {
-    selfCheck();
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java [hunk 8]
```diff
@@ -371,11 +413,100 @@ public long getOrder() {
...
+
+        flushPage(buffer, channel, currentPage * OWALPage.PAGE_SIZE);
+        channel.force(true);
+      }
+
+    } finally {
+      fileLock.unlock();
+    }
+
+    last = new OLogSequenceNumber(order, lastPosition);
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: I removed -1 initialization so only correct values can be assigned

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java (hunk 1 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: yeah, fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java (hunk 2 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java (hunk 3 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: thank you , fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java (hunk 4 above)
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: Changes to ensure backward compatibility are added.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java [hunk 9]
```diff
@@ -1,111 +1,29 @@
...
-  }
-
-  public boolean mergeWithNextPage(int position) {
-    return buffer.get(position) > 0;
-  }
 
-  public boolean isEmpty() {
-    return getFreeSpace() == MAX_ENTRY_SIZE;
-  }
+public interface OWALPage {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: No javadocs on members.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV1.java [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.CRC32;
+
+public class OLogSegmentV1 implements OLogSegment {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV1.java]: No javadocs. No override annotations on methods. `OLogSegmentV2` is package private and final while `OLogSegmentV1` is public and non-final.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV2.java [hunk 11]
```diff
@@ -0,0 +1,888 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.CRC32;
+
+import static com.orientechnologies.common.io.OIOUtils.readByteBuffer;
+import static com.orientechnologies.common.io.OIOUtils.writeByteBuffer;
+
+final class OLogSegmentV2 implements OLogSegment {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV2.java]: No javadocs. No override annotations on methods. BTW, what is our policy on javadocs for package private stuff?

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV1.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
+package com.orientechnologies.orient.core.storage.impl.local.paginated.wal;
+
+import com.orientechnologies.common.serialization.types.OIntegerSerializer;
+
+import java.nio.ByteBuffer;
+
+public class OWALPageV1 implements OWALPage {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV1.java]: No javadocs. No override annotations on methods.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV2.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+ * <li>Serialized content of the WAL record. Variable size.</li>
+ * </ol>
+ * <p>
+ * <p>
+ * Every time new record is added. Value of free space left on page is updated.
+ *
+ * @author Andrey Lomakin
+ * @since 5/8/13
+ */
+public class OWALPageV2 implements OWALPage {
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV2.java]: No override annotations on methods.

File: core/src/main/java/com/orientechnologies/common/io/OIOUtils.java [hunk 14]
```diff
@@ -339,4 +341,141 @@ public static void readFully(InputStream in, byte[] b, int off, int len) throws
...
+  /**
+   * Reads data from {@link FileChannel} to {@link ByteBuffer} starting from passed in position in {@link FileChannel}.
+   * Data will be read to {@link ByteBuffer} starting from current position till buffer limit.
+   * <p>
+   * This method is thread safe in sense that several threads can read from the same {@link FileChannel}.
+   *
+   * @param throwOnEof If parameter set to <code>true</code> then EOF exception will be thrown if end of file will be reached
+   *                   otherwise buffer will be filled with 0 bytes.
+   */
+  public static void readByteBuffer(ByteBuffer buffer, FileChannel channel, long position, boolean throwOnEof) throws IOException {
```
[the remark this answers is not in the document]
Comment (laa) [reply core/src/main/java/com/orientechnologies/common/io/OIOUtils.java]: Fixed. Thank you!

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/ODiskWriteAheadLog.java [hunk 15]
```diff
@@ -36,11 +37,13 @@
 import com.orientechnologies.orient.core.storage.impl.local.paginated.atomicoperations.OAtomicOperationMetadata;
 import com.orientechnologies.orient.core.storage.impl.local.statistic.OPerformanceStatisticManager;
 import com.orientechnologies.orient.core.storage.impl.local.statistic.OSessionStoragePerformanceStatistic;
+import sun.misc.IOUtils;
```
[the remark this answers is not in the document]
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/ODiskWriteAheadLog.java]: Fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java [hunk 16]
```diff
@@ -1,820 +1,42 @@
...
-      throw new IllegalStateException("WAL cache is not empty, we cannot verify WAL after it was started to be used");
+/**
+ * Basic interface for classes which present log segments of WAL.
+ * Main reason of creation of this interface is support of different binary formats
+ * of WAL in the same deployment.
+ * <p>
+ * To detect which version of binary format is stored we use value stored under
+ * {@link OWALPage#MAGIC_NUMBER_OFFSET}
+ */
+public interface OLogSegment extends Comparable<OLogSegment> {
```
[the remark this answers is not in the document]
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegment.java]: Fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java (hunk 9 above)
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPage.java]: Fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV1.java (hunk 10 above)
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV1.java]: Fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV2.java (hunk 11 above)
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OLogSegmentV2.java]: Well, I think they are the same as for the public. Fixed.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV1.java (hunk 12 above)
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV1.java]: Fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV2.java (hunk 13 above)
Comment (laa) [reply core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/paginated/wal/OWALPageV2.java]: Fixed

**Comment by laa** [comment]
@taburet sorry I fixed all issue but will push fixes at Monday

**Comment by laa** [comment]
@taburet pushed , could you review ?

**Pull request merged**